sale_import_base/tests/test_sale_order_import.py

```diff
@@ -0,0 +1,191 @@
+        self.assertEqual(shipping.street, "3")
+        self.assertEqual(invoicing.parent_id, customer)
+        self.assertEqual(shipping.parent_id, customer)
+
+    def test_import_existing_partner_match_external_id(self):
+        """ During import, if a partner is matched, his
+         address is updated """
+        json_import = deepcopy(self.sale_order_example_vals)
+        del json_import["address_customer"]["email"]
+        new_sale_order = self.env["sale.order"].process_json_import(json_import)
+        self.assertEqual(new_sale_order.partner_id, self.partner_thomasjean)
+
+    def test_import_existing_partner_match_email(self):
+        """ If sale channel allows it and there is no match
+        on external_id, we can match partner on email """
+        json_import = deepcopy(self.sale_order_example_vals)
+        del json_import["address_customer"]["external_id"]
+        new_sale_order = self.env["sale.order"].process_json_import(json_import)
+        self.assertEqual(new_sale_order.partner_id, self.partner_thomasjean)
+
+    def test_import_existing_partner_match_email_not_allowed(self):
+        """ DISCUSSION on en revient à créér un nouveau partner
+        1. pas match sur external_id -> 2. pas match sur email -> create. Mais
+        on risque d'avoir des conflits avec des partner dont l'email existe déjà"""
+        json_import = deepcopy(self.sale_order_example_vals)
+        del json_import["address_customer"]["external_id"]
+        new_sale_order = self.env["sale.order"].process_json_import(json_import)
+        for field in ("name", "street", "email", "city"):
+            self.assertEqual(
+                getattr(new_sale_order.partner_id, field),
+                self.addr_customer_example[field],
+            )
+
+    def test_product_missing(self):
+        """ Test product code validation effectively blocks the job """
+        json_import = deepcopy(self.sale_order_example_vals)
+        for line in json_import["lines"]:
+            line["product_code"] = "doesn't exist"
+        with self.assertRaises(ValidationError):
+            self.env["sale.order"].process_json_import(json_import)
+
+    def test_product_search(self):
+        """ Check we get the right product match on product code"""
+        json_import = deepcopy(self.sale_order_example_vals)
+        new_sale_order = self.env["sale.order"].process_json_import(json_import)
+        self.assertEqual(new_sale_order.order_line[0].product_id, self.product_order)
+        self.assertEqual(new_sale_order.order_line[1].product_id, self.product_deliver)
+
+    def test_wrong_total_amount(self):
+        """ Test the sale.exception works as intended """
+        json_import = deepcopy(self.sale_order_example_vals)
+        json_import["amount"]["amount_total"] += 500.0
+        new_sale_order = self.env["sale.order"].process_json_import(json_import)
+        with self.assertRaises(ValidationError):
+            new_sale_order._check_exception()
```

Comment:
use detect exception and check that you have the expected exception in sale.exception_ids